src/django_components/app_settings.py

```diff
@@ -93,95 +114,134 @@ class ContextBehavior(str, Enum):
+        # Fonts
+        ".eot", ".ttf", ".woff", ".otf", ".svg",
+    ],
+    forbidden_static_files=[
+        # See https://marketplace.visualstudio.com/items?itemName=junstyle.vscode-django-support
+        ".html", ".django", ".dj", ".tpl",
+        # Python files
+        ".py", ".pyc",
+    ],
+    tag_formatter="django_components.component_formatter",
+    template_cache_size=128,
+)
+# --endsnippet:defaults--
+# fmt: on
+
+
+class InternalSettings:
     @property
-    def settings(self) -> Dict:
-        return getattr(settings, "COMPONENTS", {})
+    def _settings(self) -> ComponentsSettings:
+        data = getattr(settings, "COMPONENTS", {})
+        return ComponentsSettings(**data) if not isinstance(data, ComponentsSettings) else data
 
     @property
     def AUTODISCOVER(self) -> bool:
-        return self.settings.get("autodiscover", True)
+        return default(self._settings.autodiscover, cast(bool, defaults.autodiscover))
 
     @property
-    def DIRS(self) -> List[Union[str, Tuple[str, str]]]:
-        base_dir_path = Path(settings.BASE_DIR)
-        return self.settings.get("dirs", [base_dir_path / "components"])
+    def DIRS(self) -> Sequence[Union[str, PathLike, Tuple[str, str], Tuple[str, PathLike]]]:
+        # For DIRS we use a getter, because default values uses Django settings,
+        # which may not yet be initialized at the time these settings are generated.
+        default_fn = cast(Dynamic[Sequence[Union[str, Tuple[str, str]]]], defaults.dirs)
+        default_dirs = default_fn.getter()
+        return default(self._settings.dirs, default_dirs)
 
     @property
-    def APP_DIRS(self) -> List[str]:
-        return self.settings.get("app_dirs", ["components"])
+    def APP_DIRS(self) -> Sequence[str]:
+        return default(self._settings.app_dirs, cast(List[str], defaults.app_dirs))
 
     @property
     def DYNAMIC_COMPONENT_NAME(self) -> str:
-        return self.settings.get("dynamic_component_name", "dynamic")
+        return default(self._settings.dynamic_component_name, cast(str, defaults.dynamic_component_name))
 
     @property
     def LIBRARIES(self) -> List[str]:
-        return self.settings.get("libraries", [])
+        return default(self._settings.libraries, cast(List[str], defaults.libraries))
 
     @property
     def MULTILINE_TAGS(self) -> bool:
-        return self.settings.get("multiline_tags", True)
+        return default(self._settings.multiline_tags, cast(bool, defaults.multiline_tags))
 
     @property
     def RELOAD_ON_TEMPLATE_CHANGE(self) -> bool:
-        return self.settings.get("reload_on_template_change", False)
+        return default(self._settings.reload_on_template_change, cast(bool, defaults.reload_on_template_change))
 
     @property
     def TEMPLATE_CACHE_SIZE(self) -> int:
-        return self.settings.get("template_cache_size", 128)
+        return default(self._settings.template_cache_size, cast(int, defaults.template_cache_size))
```

Comment:
(Following is outside of the scope for this MR, just sharing my thoughts)

Since the default values now live outside of `AppSettings` (`InternalSettings`), you can see that there's a lot of repetition here, and most of the properties do the same. The only difference is CONTEXT_BEHAVIOR, where we also validate the input.

This could be simplified by:
1. Moving the `CONTEXT_BEHAVIOR` validation to `ComponentsSettings`.

   At that point, all of these properties would do the same - either return value from given settings or return the default value.

2. Instead of defining all properties on `AppSettings`, we could do away with defining just a single [`__getattr__`](https://docs.python.org/3/reference/datamodel.html#object.__getattr__), that would do the same - return the key from given settings or from the defaults.

---

And when it comes to having validation defined on `ComponentsSettings`, I'd use [pydantic](https://docs.pydantic.dev/latest/) for that.

I also want to introduce pydantic as dependency down the line, because our [custom validation that I wrote](https://github.com/EmilStenstrom/django-components/blob/1fd95708d1239a46d5de19af5aa319942648f54d/src/django_components/util/validation.py) for validating Component's args and kwargs is not good enough. It doesn't handle nested TypedDict / Tuples. And I use a pattern where I define JavaScript component inputs by putting them all inside a `js` dictionary, e.g.:

```py
class TableJsProps(TypedDict):
    items: str
    items_per_page: NotRequired[str]

class TableProps(TypedDict):
    js: TableJsProps
    attrs: dict

@register("table")
class Table(Component[EmptyTuple, TableProps, ...]):
    ...
```

Which I can then use as:

```django
{% component "table"
  js:items="someJsVar"
  js:items_per_page="20"
  attrs:class="flex"
%}
```

Pydantic can handle nested structures, and it can take TypedDict and Tuples as inputs. So we wouldn't have to change the API for how the component types are defined. Only the internal implementation. Plus we could remove the whole [validation.py](https://github.com/EmilStenstrom/django-components/blob/1fd95708d1239a46d5de19af5aa319942648f54d/src/django_components/util/validation.py) file.

Reply:
Simplifying code around ComponentsSettings sounds like a good suggestion!

About adding pydantic as a dependency I'm more cautious. Remember that I would like to go for inclusion in Django one day. And Django is not typed. So adding pydantic to Django as a dependency is not going to happen. 

Overall I'm a little worried of all the complexity that types are adding across the library. I would not be comfortable touching any of the internals that deal with type validation that exists now.

Reply:
Good to know! Since I'm now thinking also about the Vue port as a standalone package that'd be build on top of django_components (let's call it django_vue), potentially the type validation could eventually live there 👀

As I'm having a look at the code, if there was the plugin support, and it would allow to hook into the flow both BEFORE and AFTER we call `get_context_data()`, then the type validation could be moved out of django_components and into django_vue.

I think there was one or two things that I wanted to experiment with that would require the use of pydantic elsewhere, but I don't remember right now what it was.

---

Although that makes me wonder, whether it makes sense, without type validation, to keep the Component generics. (TL;DR: yes, for intellisense / mypy support)

What I have in mind is this construct:

```py
Component[ArgsType, KwargsType, SlotsType, DataType, JsDataType, CssDataType]
```

My intention for the component generics was three-fold:
1. Runtime input validation
2. Static input validation (mypy) / intellisense, when using `Component.render(...)`
3. As source of truth for language server, so there could be intellisense for the django templates when using `{% component %}`.

However, with django_vue in the equation, then:
- Use case 1. would be moved to django_vue
- If I get to writing a language server (use case 3. ), then instead of writing one for Django templates, I would write one for Vue files, extending it so it supports a `<python>` (or similar) section to define the Python component.

But that means that for the use case 2., it still makes sense for the Component generics to live in django_components.